pulsar-io/couchbase/src/main/java/org/apache/pulsar/io/cb/CBAbstractSource.java

```diff
@@ -0,0 +1,314 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.io.cb;
+
+import com.couchbase.client.core.event.CouchbaseEvent;
+import com.couchbase.client.dcp.*;
+import com.couchbase.client.dcp.config.CompressionMode;
+import com.couchbase.client.dcp.message.*;
+import com.couchbase.client.dcp.transport.netty.ChannelFlowController;
+import com.couchbase.client.deps.io.netty.buffer.ByteBuf;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.functions.api.Record;
+import org.apache.pulsar.io.core.PushSource;
+import org.apache.pulsar.io.core.SourceContext;
+import rx.CompletableSubscriber;
+import rx.Subscription;
+
+import java.io.Closeable;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.couchbase.client.dcp.config.DcpControl.Names.ENABLE_NOOP;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+@Slf4j
+public abstract class CBAbstractSource<T> extends PushSource<T> {
+  protected Thread thread = null;
+
+  private CBSourceConfig cbSourceConfig;
+
+  private Client client;
+
+  protected final Thread.UncaughtExceptionHandler handler = new Thread.UncaughtExceptionHandler() {
+    @Override
+    public void uncaughtException(Thread thread, Throwable ex) {
+      log.error("{} has an error", thread.getName(), ex);
+    }
+  };
+
+  private ConcurrentMap<Short, DcpSnapshotMarkerRequestMeta> snapshots = new ConcurrentHashMap();
+
+  /**
+   * Open connector with configuration
+   *
+   * @param config        initialization config
+   * @param sourceContext
+   * @throws Exception IO type exceptions when opening a connector
+   */
+  @Override
+  public void open(Map<String, Object> config, SourceContext sourceContext) throws Exception {
+    log.info("opening connection to couchbase.");
+
+    cbSourceConfig = CBSourceConfig.load(config);
+//        cbSourceConfig.validate();
```

Comment:
Could remove unused code, and whether the required parameters need to be verified like cbSourceConfig.validate() or use Preconditions.checkNotNull()